---
title: "DG_GRN_inferfence"
output: html_document
date: "`r Sys.Date()`"
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(echo = TRUE)

#Set random seed
set.seed(123)
RNGkind("L'Ecuyer-CMRG")

##Load required packages
library(msigdbr)
library(dplyr)
library(tidyr)
library(stringr)
library(Matrix)
library(EnsDb.Mmusculus.v79)
library(arrow)
library(SingleCellExperiment)
library(readr)
library(GenomicRanges)
library(ComplexHeatmap)
library(patchwork)
library(Seurat)
library(gtools)
library(Signac)
library(ggtext)
library(glue)
library(patchwork)
library(ggplot2)
library(scCustomize)
library(future)
library(ggdark)
library(tidyverse)
library(ggrepel)
library(clusterProfiler)
library(enrichplot)
library(fgsea)
library(edgeR)
library(RcppML)
library(scDblFinder)
library(wrapr)
library(transformGamPoi)
library(ggprism)
library(EasyMultiome)
library(igraph)
library(visNetwork)
library(htmlwidgets)
#library(consensusClustR)
library(ggtext)
library(glue)
library(DOSE)

 highlight = function(x, pat, color="black", family="") {
    ifelse(grepl(pat, x), glue("<b style='font-family:{family}; color:{color}'>{x}</b>"), x)
 }
 
httr::set_config(httr::config(ssl_verifypeer = FALSE))

##Start Python
library(cluster)
library(reticulate)
path_to_python <- "/home/andy/anaconda3/bin/python"
use_python(path_to_python)

##Set Settings
options(future.globals.maxSize = 10000 * 1024^2)

# Some relative paths
cellRanger = "../../../data/processed/cellrangerOutput/"
clusters = "../../../results/multiome/seurat_analysis/clusters/"
plots = "../../../results/multiome/seurat_analysis/plots/"
objects =  "../../../data/multiome/seurat_analysis/objects/"

# Colour palette
cb_pallette = c("#777a78", "#E4E1E3FF", "#F6222EFF", "#FE00FAFF", "#16FF32FF", "#3283FEFF", "#FEAF16FF", "#B00068FF", "#1CFFCEFF", "#90AD1CFF", "#2ED9FFFF", "#DEA0FDFF", "#AA0DFEFF", "#F8A19FFF", "#325A9BFF", "#C4451CFF", "#1C8356FF", "#85660DFF", "#B10DA1FF", "#FBE426FF", "#1CBE4FFF", "#FA0087FF", "#FC1CBFFF", "#F7E1A0FF", "#C075A6FF", "#782AB6FF", "#AAF400FF", "#BDCDFFFF", "#f2b672", "#B5EFB5FF", "#f75c02", "#5c0404", "#160459", "#FFFFCC", "#DF65B0", "#B3E2CD", "#FBB4AE", "#1F78B4", "#BDC9E1","#FDAE61", "#FFFF99","#02818A", "#F1B6DA","#DFC27D")
```

### First set up some gene-sets for later enrichment analysis of GRNs

```{r}
load(paste0(objects, "/geneSets/GOterms.rds"))
Chen_L_LTPgenes <- readxl::read_excel(paste0(objects, "geneSets/Chen_L-LTPgenes.XLSX"), 
    sheet = "Table S1", skip = 3)

Chen_L_LTPgenes = Chen_L_LTPgenes[order(Chen_L_LTPgenes$log2FC, decreasing = TRUE), ]
GOterms[["Chen_LTPUP_30'"]] = Chen_L_LTPgenes[Chen_L_LTPgenes$FDR < 0.01 & Chen_L_LTPgenes$`Time point` == "30'" & Chen_L_LTPgenes$log2FC > 0.75,]$...1
GOterms[["Chen_LTPUP_60'"]] = Chen_L_LTPgenes[Chen_L_LTPgenes$FDR < 0.01 & Chen_L_LTPgenes$`Time point` == "60'" & Chen_L_LTPgenes$log2FC > 0.75,]$...1
GOterms[["Chen_LTPUP_120'"]] = Chen_L_LTPgenes[Chen_L_LTPgenes$FDR < 0.01 & Chen_L_LTPgenes$`Time point` == "120'" & Chen_L_LTPgenes$log2FC > 0.75,]$...1

Chen_L_LTPgenes = Chen_L_LTPgenes[order(Chen_L_LTPgenes$log2FC, decreasing = FALSE), ]
GOterms[["Chen_LTPDown_30'"]] = Chen_L_LTPgenes[Chen_L_LTPgenes$FDR < 0.05 & Chen_L_LTPgenes$`Time point` == "30'" & Chen_L_LTPgenes$log2FC < -0.5,]$...1
GOterms[["Chen_LTPDown_60'"]] = Chen_L_LTPgenes[Chen_L_LTPgenes$FDR < 0.05 & Chen_L_LTPgenes$`Time point` == "60'" & Chen_L_LTPgenes$log2FC < -0.5,]$...1
GOterms[["Chen_LTPDown_120'"]] = Chen_L_LTPgenes[Chen_L_LTPgenes$FDR < 0.05 & Chen_L_LTPgenes$`Time point` == "120'" & Chen_L_LTPgenes$log2FC < -0.5,]$...1
```

## Infer the GRNs

Load PRINT footprinting results

```{r}
Cluster = "CA1"
load(paste0(clusters,Cluster, "/clusters/objects/regulon.rds"))
boundRegions = readRDS("../../PRINT/CA1/boundRegions.rds") #foot printing PRINT footprinting
source("GRN_inference.R")

#Select the most variable TFs with bound regions detected
boundRegions = boundRegions[sapply(boundRegions, \(TF) !is.null(TF))]
TFnames = setNames(names(boundRegions), names(boundRegions))
TFnames = TFnames[TFnames %in% rownames(variability[variability$p_value<0.05,])]
TFnames = c(TFnames, setNames(c("Mef2a", "Mef2b", "Mef2c"), c("Mef2a", "Mef2b", "Mef2c")))

# RUn the inference
Res=FindTFregulons(TFnames, boundRegions=boundRegions, chromvarMat = chromvarMat, chromvar.obj = chromvar.obj, motifMat = MotifMat, rnaMatrix = rnaMatrix, annotation = annotation, peakMatrix = peakMatrix, genesets = GOterms)

#Add to Multiome object
Multiome = readRDS(paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))
Multiome[["specific_peaks"]]@misc = Res
saveRDS(Multiome, paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))
saveRDS(Res, paste0(clusters,Cluster, "/clusters/objects/Regulons.rds"))
```

## Make some barplots of GRN GSEA enrichments
```{r}
Cluster="CA1"
Multiome = readRDS(paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))

barplot(Multiome[["specific_peaks"]]@misc$GRNs$Fos$pos$GSEA, showCategory = c("taxis", "synaptic signaling", "neuron projection morphogenesis", "neuron development", "synapse organization", "cell junction organization"), x = "GeneRatio") + viridis::scale_fill_viridis(direction = -1) + theme_ggprism_mod() + theme(axis.text.y=element_markdown(size = 18)) + scale_y_discrete(labels= function(x) highlight(x, "behavior|synaptic signaling|neuron projection morphogenesis", "#46B9DB")) + xlab("Ratio of GRN Genes in Set") + ggtitle(expression(paste("Gene-Set Enrichment of ", italic("Fos"), " Positive GRN in CA1"))) 
ggsave(file = paste0(clusters,Cluster,"/plots/CA1_FosGRNenrichment.jpg"), device = "jpeg", width = 10, dpi = 600, height = 7, bg = "white")
```

## Conduct DiffExpAnalysis To See Which Sets of Genes are Unusually Highly Expressed by IEG-expressing or accessible cells

```{r}
Cluster = "ExNeu"
Multiome = readRDS(paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))

Multiome[["regulons"]] = CreateAssay5Object(data=Multiome[["specific_peaks"]]@misc$TFactivities)
YCvsYN_regulons = FindMarkers(Multiome, assay = "regulons", group.by = "ID", ident.1 = "YC", ident.2="YN")
DotPlot(Multiome, features = c("Fos", "Fosb", "Junb", "Fosl2"), assay = "regulons", group.by = "ID", scale.min = 0, scale = F)
DotPlot(Multiome, features = c("Mef2a", "Mef2b", "Mef2c"), assay = "regulons", group.by = "ID", scale.min = 0, scale = F)

Multiome$Engram = InductionPlot(Multiome, Var = c("Fos", "Fosb", "Junb", "Fosl2"), Group = "Memory",Group2 = "Age", ylab = "Proportion of Engram Neurons", facet = "clusters", xlab = "Memory State", ncol = 2, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), Thresh = 4.25,  PrintThreshPlot = F, assay = "regulons",  nbin = 20, ctrl = 15, Return_Assignments = T) 
ggsave(file = paste0(clusters,Cluster, "/plots/CA1_Engram_Induction_Clusters.png"), device = "png", width = 7, height = 6, bg = "White")

Multiome$Engram = InductionPlot(Multiome, Var = c("Fos", "Fosb", "Junb", "Fosl2"), Group = "Memory",Group2 = "Age", ylab = "Proportion of Engram Neurons", facet = "clusters", xlab = "Memory State", ncol = 1, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), Thresh = 4.25,  PrintThreshPlot = F, assay = "regulons",  nbin = 20, ctrl = 15, Return_Assignments = T) 
ggsave(file = paste0(clusters,Cluster, "/plots/CA1_Engram_Induction_Clusters_long.png"), device = "png", width = 4, height = 12, bg = "White")

Results = DoIntegratedAnalysis(MULTIobj = Multiome, MULTIcontrasts = "GroupVarset-GroupNo", Factor = "Engram", contrastNames = "EngramRegulons",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)

Results = DoIntegratedAnalysis(MULTIobj = subset(Multiome, idents = "ExNeu.CA1.Dor.Potentiated"), MULTIcontrasts = "GroupVarset-GroupNo", Factor = "Engram", contrastNames = "EngramRegulons_dpCA1",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)
```

## Plot ridgeplot of GSEA results

```{r}
gsea = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulons_dpCA1_GSEA.csv"))
ranks = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulons_dpCA1_EdgeR.csv"))
ranks = ranks[order(ranks$logFC, decreasing = T),]
ranks = setNames(ranks$logFC, ranks$X)
source("../../../../../apps/EasyMultiome/R/Plotting.R")
ridgeplot.FGSEA(gsea, ranks, showCategory = c("cell adhesion", "cell-cell adhesion via plasma-membrane adhesion molecules", "nervous system development"), title = "GO Term Differences Between dpCA1 Engram Neurons and Other dpCA1 Neurons", 
                xhighlight = "cell adhesion", xhighlight_col = "#46B9DB") 
ggsave(file = paste0(clusters,Cluster,"/plots/dpCA1_Engram_GSEA_RidgePlot.jpg"), device = "jpeg", width = 12, dpi = 600, height = 8, bg = "white")
```

## Now investigate cells with repressed engram regulons

```{r}
save = Multiome[["regulons"]]["data"]
Multiome[["regulons"]]["data"] = 0 - Multiome[["regulons"]]["data"]
Multiome$EngramRegulonsRepressed = InductionPlot(Multiome, Var =c("Fos", "Fosb", "Junb", "Fosl2") , Group = "Memory", Group2 = "Age", ylab = paste0("Proportion of Neurons Repressing IEG Motifs"), xlab = "Memory State", Thresh = 3.25, Return_Assignments = TRUE, All = T, assay = "regulons", nbin = 15, ctrl = 15, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), ncol = 2, PrintThreshPlot = F, facet = "clusters")
ggsave(file = paste0(clusters,Cluster, "/clusters/plots/CA1_IEGregulon_Repression.png"), device = "png", dpi = 600, width = 6, height = 6, bg = "White")
Multiome$EngramRegulonsRepressed = InductionPlot(Multiome, Var =c("Fos", "Fosb", "Junb", "Fosl2") , Group = "Memory", Group2 = "Age", ylab = paste0("Proportion of Neurons Repressing IEG Motifs"), xlab = "Memory State", Thresh = 3.25, Return_Assignments = TRUE, All = T, assay = "regulons", nbin = 15, ctrl = 15, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), ncol = 1, PrintThreshPlot = F, facet = "clusters")
ggsave(file = paste0(clusters,Cluster, "/clusters/plots/CA1_IEGregulon_Repression_long.png"), device = "png", dpi = 600, width = 6, height = 6, bg = "White")
Multiome[["regulons"]]["data"] = save

Results = DoIntegratedAnalysis(MULTIobj = Multiome, MULTIcontrasts = "GroupVarset-GroupNo", Factor = "EngramRegulonsRepressed", contrastNames = "EngramRegulonsRepressed",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)

Results = DoIntegratedAnalysis(MULTIobj = subset(Multiome, idents = "ExNeu.CA1.Ven"), MULTIcontrasts = "GroupVarset-GroupNo", Factor = "EngramRegulonsRepressed", contrastNames = "EngramRegulonsRepressed_Ven",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)
```

## Plot ridgeplot of GSEA results

```{r}
gsea = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulonsRepressed_Ven_GSEA_Full.csv"))
ranks = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulonsRepressed_Ven_EdgeR.csv"))
ranks = ranks[order(ranks$logFC, decreasing = T),]
ranks = setNames(ranks$logFC, ranks$X)
source("../../../../../apps/EasyMultiome/R/Plotting.R")
ridgeplot.FGSEA(gsea, ranks, showCategory = c("synaptic signaling", "behavior", "receptor localization to synapse", "synapse assembly", "cell-cell adhesion", "neurotransmitter transport", "learning or memory", "receptor clustering", "dendrite morphogenesis"), title = "GO Term Differences Between IEG-Motif Repressed vCA1 Neurons and Other vCA1 Neurons", 
                xhighlight = "synaptic signaling|behavior|synapse assembly|cell-cell adhesion", xhighlight_col = "#46B9DB")
ggsave(file = paste0(clusters,Cluster,"/plots/vCA1_IEGmotifRepressed_GSEA_RidgePlot.jpg"), device = "jpeg", width = 13, dpi = 600, height = 9, bg = "white")

saveRDS(Multiome, file = paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))
```

## Regulon activity heatmap

```{r}
Cluster="ExNeu"
#Multiome = readRDS(paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))
#Motif activity heatmap by ID

Multiome$clusters = Idents(Multiome)
Idents(Multiome) = "ID"
DS = subset(Multiome, downsample = 350)
Idents(Multiome) = "clusters"
IDs = factor(DS$ID, levels = c("YN", "YC", "ON", "OC"))
DefaultAssay(DS) = "regulons"

matrix =  t(scale(t(DS@assays$regulons["data"])))

# what's the value range in the matrix
quantiles = quantile(matrix, c(0.00005, 0.9998))
## make the black color map to 0. the yellow map to highest and the purle map to the lowest
col_fun = circlize::colorRamp2(c(quantiles[1], 0, quantiles[2]), c("#FF00FF", "black", "#FFFF00"))

#Consensus hierachical clustering

png(file=paste0(clusters,Cluster, "/plots/CA1_regulon_HeatmapID.png"), width = 1050, height = 3750)
dend2 = cluster_within_group(matrix, IDs)
region_ht = Heatmap(matrix, cluster_columns = dend2, col = col_fun, row_names_gp = gpar(fontsize = 12.5),show_row_names = T, show_column_names = FALSE, row_title = "GRN Activity", column_title = "GRN Activity by Age and Memory State", column_title_gp = gpar(fontsize = 35),
             top_annotation = HeatmapAnnotation(ID = IDs, 
                                col = list(ID = c("YN" = "royalblue1", "YC" = "blue", "ON" = "tomato", "OC" = "red")),
                                                annotation_legend_param = list(ID = list(
                                                  grid_height = unit(5, "cm"), 
                                                  title = "Group ID",
                                                  title_gp = gpar(fontsize = 20, fontface = "bold"),
                                                  labels_gp = gpar(fontsize = 20),
                                                  title_position = "lefttop-rot")),
                                                annotation_name_gp= gpar(fontsize = 30), show_annotation_name = FALSE
                                                  ), 
             column_split = 4, column_gap = unit(2, "mm"), row_gap = unit(2, "mm"), km = 7, row_km_repeats = 1000, 
             heatmap_legend_param = list(legend_height = unit(15, "cm"), 
                                         title = "Motif Accessibility",
                                         title_position = "lefttop-rot",
                                         labels_gp = gpar(fontsize = 20),
                                         title_gp = gpar(fontsize = 20, fontface = "bold")
                                         #labels = c("very low", "low", "medium", "high")
                                         )
             
             # annotation_legend_param = list(legend_height = unit(15, "cm"),
             #                                title = "Group")
)
region_ht = draw(region_ht)
dev.off()
b=row_order(region_ht)

bnames = sapply(b, function(x) unlist(rownames(DS@assays$regulons["data"])[match(rownames(matrix)[unlist(x)],rownames(DS@assays$regulons["data"]))]) ##Get TF names for third cluster
)
```

## Zoom in on IEGs
```{r}
matrix = matrix[b[[6]],]
# what's the value range in the matrix
quantiles = quantile(matrix, c(0.00005, 0.9998))
## make the black color map to 0. the yellow map to highest and the purle map to the lowest
col_fun = circlize::colorRamp2(c(quantiles[1], 0, quantiles[2]), c("#FF00FF", "black", "#FFFF00"))

png(file=paste0(clusters,Cluster, "/plots/CA1_regulon_HeatmapID_IEGsOnly.png"), width = 3050, height = 1200)
region_ht = Heatmap(matrix, cluster_columns = dend2, cluster_rows = F, col = col_fun, row_names_gp = gpar(fontsize = 40),show_row_names = T, show_column_names = FALSE, column_title_gp = gpar(fontsize = 35),
             top_annotation = HeatmapAnnotation(ID = IDs, 
                                col = list(ID = c("YN" = "royalblue1", "YC" = "blue", "ON" = "tomato", "OC" = "red")),
                                                annotation_legend_param = list(ID = list(
                                                  grid_height = unit(5, "cm"), 
                                                  title = "Group ID",
                                                  title_gp = gpar(fontsize = 20, fontface = "bold"),
                                                  labels_gp = gpar(fontsize = 20),
                                                  title_position = "lefttop-rot")),
                                                annotation_name_gp= gpar(fontsize = 30), show_annotation_name = FALSE
                                                  ), 
             column_split = 4, column_gap = unit(2, "mm"), row_gap = unit(2, "mm"), km = 1
             
             # annotation_legend_param = list(legend_height = unit(15, "cm"),
             #                                title = "Group")
)
region_ht = draw(region_ht)
dev.off()
```

## Induction plot of the proportion of cells high accessible or repressed for IEG regulons

```{r}
Multiome$EngramRegulons = InductionPlot(Multiome, Var = bnames[["6"]], Group = "Memory",Group2 = "Age", ylab = "Proportion of Engram Neurons", facet = "clusters", xlab = "Memory State", ncol = 2, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), Thresh = 2.75,  PrintThreshPlot = F, assay = "regulons",  nbin = 20, ctrl = 15, Return_Assignments = T) 
ggsave(file = paste0(clusters,Cluster, "/plots/CA1_Engram_Induction_Clusters_kmeans.png"), device = "png", width = 7, height = 7, bg = "White")

Multiome$EngramRegulons = InductionPlot(Multiome, Var = bnames[["6"]], Group = "Memory",Group2 = "Age", ylab = "Proportion of Engram Neurons", facet = "clusters", xlab = "Memory State", ncol = 1, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), Thresh = 2.75,  PrintThreshPlot = F, assay = "regulons",  nbin = 20, ctrl = 15, Return_Assignments = T) 
ggsave(file = paste0(clusters,Cluster, "/plots/CA1_Engram_Induction_Clusters_long_kmeans.png"), device = "png", width = 4, height = 12, bg = "White")
```

## Now repeat differential expression analysis using the kmeans cluster of IEGs, rather than a manually selected set

```{r}
# Add the IEG GRN genes as gene sets
IEGgrns = lapply(setNames(bnames[["6"]], bnames[["6"]]), \(TF) unique(Multiome[["specific_peaks"]]@misc$GRNs[[TF]]$pos$links$gene) )
GOterms = c(GOterms, IEGgrns)

Results = DoIntegratedAnalysis(MULTIobj = Multiome, MULTIcontrasts = "GroupVarset-GroupNo", Factor = "Engram", contrastNames = "EngramRegulons_kmeans",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)

Results = DoIntegratedAnalysis(MULTIobj = subset(Multiome, idents = "ExNeu.CA1.Dor.Potentiated"), MULTIcontrasts = "GroupVarset-GroupNo", Factor = "Engram", contrastNames = "EngramRegulons_dpCA1_kmeans",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=F, minGenesetSize=10, maxGenesetSize=Inf)
```

## Plot ridgeplot of GSEA results

```{r}
gsea = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulons_dpCA1_kmeans_GSEA.csv"))
ranks = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulons_dpCA1_kmeans_EdgeR.csv"))
ranks = ranks[order(ranks$logFC, decreasing = T),]
ranks = setNames(ranks$logFC, ranks$X)
source("../../../../../apps/EasyMultiome/R/Plotting.R")
ridgeplot.FGSEA(gsea, ranks, showCategory = c("cell adhesion", "cell-cell adhesion via plasma-membrane adhesion molecules", "nervous system development"), title = "GO Term Differences Between dpCA1 Engram Neurons and Other dpCA1 Neurons", 
                xhighlight = "cell adhesion", xhighlight_col = "#46B9DB") 
ggsave(file = paste0(clusters,Cluster,"/plots/dpCA1_Engram_kmeans_GSEA_RidgePlot.jpg"), device = "jpeg", width = 12, dpi = 600, height = 8, bg = "white")
```

## Now for IEG motif repression 

```{r}
save = Multiome[["regulons"]]["data"]
Multiome[["regulons"]]["data"] = 0 - Multiome[["regulons"]]["data"]
Multiome$EngramRegulonsRepressed = InductionPlot(Multiome, Var =bnames[["6"]] , Group = "Memory", Group2 = "Age", ylab = paste0("Proportion of Neurons With Highly Repressed IEG Motifs"), xlab = "Memory State", Thresh = 2, Return_Assignments = TRUE, All = T, assay = "regulons", nbin = 15, ctrl = 15, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), ncol = 2, PrintThreshPlot = F, facet = "clusters")
ggsave(file = paste0(clusters,Cluster, "/clusters/plots/CA1_IEGregulon_Repression_kmeans.png"), device = "png", dpi = 600, width = 7, height = 7, bg = "White")
Multiome$EngramRegulonsRepressed = InductionPlot(Multiome, Var =bnames[["6"]] , Group = "Memory", Group2 = "Age", ylab = paste0("Proportion of Neurons With Highly Repressed IEG Motifs"), xlab = "Memory State", Thresh = 2, Return_Assignments = TRUE, All = T, assay = "regulons", nbin = 15, ctrl = 15, GroupOrder = c("Naive", "CFC"), Group2Order = c("Young", "Aged"), lineCol = c("Blue", "Red"), ncol = 1, PrintThreshPlot = F, facet = "clusters")
ggsave(file = paste0(clusters,Cluster, "/clusters/plots/CA1_IEGregulon_Repression_long_kmeans.png"), device = "png", dpi = 600, width = 4, height = 10, bg = "White")
Multiome[["regulons"]]["data"] = save

Results = DoIntegratedAnalysis(MULTIobj = Multiome, MULTIcontrasts = "GroupVarset-GroupNo", Factor = "EngramRegulonsRepressed", contrastNames = "EngramRegulonsRepressed_kmeans",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)

Results = DoIntegratedAnalysis(MULTIobj = subset(Multiome, idents = "ExNeu.CA1.Ven"), MULTIcontrasts = "GroupVarset-GroupNo", Factor = "EngramRegulonsRepressed", contrastNames = "EngramRegulonsRepressed_Ven_kmeans",Genesets = GOterms, geneThreshold = 0.01, regionThreshold = 0.05, save_to = file.path(clusters,Cluster, "/"), DARthresh = list(0.05, 0.3), pseudobulk = F, addMotifs = F, ResultsTable = FALSE, CollapsePathways=T, minGenesetSize=10, maxGenesetSize=Inf)
```

## Plot ridgeplot of GSEA results

```{r}
gsea = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulonsRepressed_Ven_kmeans_GSEA_Full.csv"))
ranks = read.csv(file.path(clusters,Cluster, "/DifferentialExpression/EngramRegulonsRepressed_Ven_kmeans_EdgeR.csv"))
ranks = ranks[order(ranks$logFC, decreasing = T),]
ranks = setNames(ranks$logFC, ranks$X)
source("../../../../../apps/EasyMultiome/R/Plotting.R")
ridgeplot.FGSEA(gsea, ranks, showCategory = c("synaptic signaling", "behavior", "receptor localization to synapse", "synapse assembly", "cell-cell adhesion", "neurotransmitter transport", "learning or memory","neuron differentiation","membrane lipid catabolic process","	
amide biosynthetic process", "receptor clustering", "dendrite morphogenesis"), title = "GO Term Differences Between IEG-Motif Repressed vCA1 Neurons and Other vCA1 Neurons", 
                xhighlight = "synaptic signaling|behavior|synapse assembly|cell-cell adhesion", xhighlight_col = "#46B9DB")
ggsave(file = paste0(clusters,Cluster,"/plots/vCA1_IEGmotifRepressed_kmeans_GSEA_RidgePlot.jpg"), device = "jpeg", width = 13, dpi = 600, height = 9, bg = "white")

saveRDS(Multiome, file = paste0(clusters,Cluster, "/clusters/objects/CA1.rds"))
```

### Now lets plot the GRNs of "Fos", "Fosb", "Junb", and "Fosl2" (IEGs)

```{r}
Cluster = "CA1"
Multiome = readRDS(paste0(clusters,Cluster, "/objects/Multiome.rds"))

TFoI = c("Fos", "Fosb", "Junb", "Fosl2")
#Get region-geneLinks
Res = as.data.frame(do.call(rbind, purrr::map(Multiome[["specific_peaks"]]@misc$GRNs, \(x) x$pos[[1]])))
Res = Res[Res$TF %in% TFoI,]

#Only keep most significant positive associations
Res = Res[Res$TFtoGeneR > 0 & Res$PeaktoGeneR > 0 & Res$TFtoPeakR > 0,]
TFtoREG = Res[,c(12, 5, 11)]
colnames(TFtoREG) = c("from", "to", "beta")
TFtoREG$beta= as.numeric(TFtoREG$beta )
TFtoREG$beta = TFtoREG$beta/ max(TFtoREG$beta)
TFtoREG = TFtoREG[!duplicated(TFtoREG),]

REGtoGene = Res[,c(5, 1, 8)]
colnames(REGtoGene) = c("from", "to", "beta")
REGtoGene = REGtoGene[!duplicated(REGtoGene),]
REGtoGene$beta= as.numeric(REGtoGene$beta )
REGtoGene$beta = REGtoGene$beta/ max(REGtoGene$beta)
GraphDF = rbind(TFtoREG, REGtoGene)

#Add node data
nodeinfo = data.frame(names = c(unique(GraphDF$from), unique(REGtoGene$to)), nodeType = rep(c("TF", "Region", "Gene"), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))), 
                      color = rep(c("Red", "Orange", "Yellow"), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
                      label = c(unique(TFtoREG$from), rep(NA, length(unique(REGtoGene$from))), unique(REGtoGene$to)))

nodeinfo = nodeinfo[!duplicated(nodeinfo$names),]

#Make igraph object
Net = graph_from_data_frame(GraphDF, directed = TRUE, vertices = nodeinfo)

#Alter layout
layout <- layout_nicely(Net)

#Initial plot
plot(Net, layout = layout, 
     vertex.size = rep(c(10, 2, 4), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     vertex.label.cex = rep(c(1, 0, 0.6), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))), vertex.shape = rep(c("sphere", "csquare", "circle"), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     edge.arrow.size = 0.25, edge.width = E(Net)$beta,
     vertex.label.color = "black",
     )

#Interactive plot to change layout if desired
tkid = tkplot(Net,layout = layout, 
     vertex.size = rep(c(10, 2, 4), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     vertex.label.cex = rep(c(1, 0, 0.6), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     vertex.shape = rep(c("sphere", "csquare", "circle"), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     edge.arrow.size = 0.5, edge.width = E(Net)$beta,
     vertex.label.color = "black") #tkid is the id of the tkplot that will open
```

### Now lets save the network plot
```{r}
layout <- tkplot.getcoords(tkid) # grab the coordinates from tkplot (while still running)
tk_close(tkid, window.close = T)

jpeg(file= paste0(clusters,Cluster, "/plots/EngramGRNs.jpg"), width = 5000, height = 5000, quality = 90, res = 300)
plot(Net, layout = layout, 
     vertex.size = rep(c(10, 2, 4), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     vertex.label.cex = rep(c(1, 0, 0.7), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     vertex.shape = rep(c("circle", "csquare", "circle"), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))),
     edge.arrow.size = 0.5, edge.width = E(Net)$beta,
     vertex.label.color = "black", vertex.label.family = "sans",
     vertex.label.font = rep(c(4, 1, 3), 
                                                  times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to)))))
dev.off()

#Also as an interactive network image - change properties on Net to change plotting parameters here
V(Net)$size = rep(c(20, 2, 4), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to))))
V(Net)$label.cex = rep(c(2, 0, 0.6), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to))))
V(Net)$shape = rep(c("sphere", "csquare", "circle"), times = c(length(unique(TFtoREG$from)), length(unique(REGtoGene$from)), length(unique(REGtoGene$to))))

E(Net)$arrow.size = 0.25
E(Net)$width = E(Net)$beta

plot(visIgraph(Net))
data <- toVisNetworkData(Net)
visNetwork(nodes = data$nodes, edges = data$edges, height = "1000px")

dat = visNetwork(nodes = data$nodes, edges = data$edges, height = "2000px") %>% visIgraphLayout()
saveWidget(dat, file = paste0(clusters,Cluster, "/plots/EngramGRNs.html"))
```
